src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -286,52 +517,90 @@ static size_t uct_cuda_copy_md_get_total_device_mem(CUdevice cuda_device)
     return 1; /* return 1 byte to avoid division by zero */
 }
 
+static void uct_cuda_copy_sync_memops(uct_cuda_copy_md_t *md,
+                                      const void *address)
+{
+#if HAVE_CUDA_FABRIC
+    ucs_status_t status;
+    if (!md->sync_memops_set) {
+        /* Synchronize future DMA operations for all memory types */
+        status = UCT_CUDADRV_FUNC_LOG_WARN(cuCtxSetFlags(CU_CTX_SYNC_MEMOPS));
+        if (status == UCS_OK) {
+            md->sync_memops_set = 1;
+        }
+    }
+#else
+    unsigned value = 1;
+    /* Synchronize for DMA for legacy memory types*/
+    UCT_CUDADRV_FUNC_LOG_WARN(
+            cuPointerSetAttribute(&value, CU_POINTER_ATTRIBUTE_SYNC_MEMOPS,
+                                  (CUdeviceptr)address));
+#endif
+
+}
+
 static ucs_status_t
 uct_cuda_copy_md_query_attributes(uct_cuda_copy_md_t *md, const void *address,
                                   size_t length, ucs_memory_info_t *mem_info)
 {
 #define UCT_CUDA_MEM_QUERY_NUM_ATTRS 4
-    CUmemorytype cuda_mem_mype = (CUmemorytype)0;
+    CUmemorytype cuda_mem_type = CU_MEMORYTYPE_HOST;
     uint32_t is_managed        = 0;
     CUdevice cuda_device       = -1;
     CUcontext cuda_mem_ctx     = NULL;
     CUpointer_attribute attr_type[UCT_CUDA_MEM_QUERY_NUM_ATTRS];
     void *attr_data[UCT_CUDA_MEM_QUERY_NUM_ATTRS];
+    ucs_memory_type_t vmm_mem_type;
     CUdeviceptr base_address;
     size_t alloc_length;
     size_t total_bytes;
     int32_t pref_loc;
+    unsigned is_vmm;
     CUresult cu_err;
+    ucs_status_t status;
 
     attr_type[0] = CU_POINTER_ATTRIBUTE_MEMORY_TYPE;
-    attr_data[0] = &cuda_mem_mype;
+    attr_data[0] = &cuda_mem_type;
     attr_type[1] = CU_POINTER_ATTRIBUTE_IS_MANAGED;
     attr_data[1] = &is_managed;
     attr_type[2] = CU_POINTER_ATTRIBUTE_DEVICE_ORDINAL;
     attr_data[2] = &cuda_device;
     attr_type[3] = CU_POINTER_ATTRIBUTE_CONTEXT;
     attr_data[3] = &cuda_mem_ctx;
 
-    cu_err = cuPointerGetAttributes(ucs_static_array_size(attr_data), attr_type,
-                                    attr_data, (CUdeviceptr)address);
-    if ((cu_err != CUDA_SUCCESS) || (cuda_mem_mype != CU_MEMORYTYPE_DEVICE)) {
+    status = UCT_CUDADRV_FUNC_LOG_ERR(
+            cuPointerGetAttributes(ucs_static_array_size(attr_data), attr_type,
+                                   attr_data, (CUdeviceptr)address));
+    if (status != UCS_OK) {
+        /* pointer not recognized */
+        return status;
+    }
+
+    is_vmm = uct_cuda_copy_detect_vmm((void*)address, &vmm_mem_type);
```

Comment:
as discussed, try the following approach:
 detect_vmm first, so if the memory is VMM take the properties from the internal call to cuMemGetAllocationPropertiesFromHandle, instead of also using cuPointerGetAttributes

Reply:
@yosefe Looked into this more. Today we query the following:
```C
attr_type[0] = CU_POINTER_ATTRIBUTE_MEMORY_TYPE;
attr_data[0] = &cuda_mem_type;
attr_type[1] = CU_POINTER_ATTRIBUTE_IS_MANAGED;
attr_data[1] = &is_managed;
attr_type[2] = CU_POINTER_ATTRIBUTE_DEVICE_ORDINAL;
attr_data[2] = &cuda_device;
```
if we detected if memory is vmm, we could get away without querying for any of the attributes because currently managed memory isn't supported and we get everything else from allocation properties. However, before we check if memory is vmm, we do need to ensure that CUDA recognizes the memory as CU_MEMORYTYPE_DEVICE (fabric pinned) or CU_MEMORYTYPE_HOST (fabric EGM). The difference in pointer querying memory type and querying more attributes (like managed/context/ordinal) is supposed to be insignificant. On the other hand, if the memory happens to not be VMM then we have to additionally query for other pointer attributes and that would effectively double query time for non-VMM cases. This being the case, I don't think checking for vmm first is justified. What are your thoughts?

Reply:
IMO from code design perspective, it makes more sense that if we detect that memory is VMM, we take the properties from VMM flow, otherwise we fallback to the existing cuPointerQuery method.